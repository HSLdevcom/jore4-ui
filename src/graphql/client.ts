import {
  ApolloClient,
  concat,
  HttpLink,
  InMemoryCache,
  split,
} from '@apollo/client';
import { WebSocketLink } from '@apollo/client/link/ws';
import { getMainDefinition } from '@apollo/client/utilities';
import { mapHttpToWs } from '../utils';
import { authRoleMiddleware } from './auth';

export const createGraphqlClient = () => {
  const graphqlServerUrl = '/api/graphql/v1/graphql';

  const httpLink = new HttpLink({
    uri: graphqlServerUrl,
  });

  const apolloLink = concat(authRoleMiddleware, httpLink);

  // because next.js might run this on server-side and websockets aren't
  // supported there, we have to check if we are on browser before
  // initializing WebSocket link
  const wsLink = process.browser
    ? new WebSocketLink({
        // WebSocketLink doesn't work with relative url's, so we have to
        // turn relative url into absolute.
        uri: `${mapHttpToWs(window.location.origin)}${graphqlServerUrl}`,
        options: {
          reconnect: true,
        },
      })
    : undefined;

  const link = process.browser
    ? split(
        ({ query }) => {
          const definition = getMainDefinition(query);
          const isSubscription =
            definition.kind === 'OperationDefinition' &&
            definition.operation === 'subscription';
          return isSubscription;
        },
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        wsLink!,
        apolloLink,
      )
    : apolloLink;

  const cache = new InMemoryCache({
    typePolicies: {
      route_route: {
        keyFields: ['route_id'],
      },
      service_pattern_scheduled_stop_point: {
        keyFields: ['scheduled_stop_point_id'],
      },
    },
  });

  const client = new ApolloClient({
    link,
    cache,
    defaultOptions: {
      watchQuery: {
        // Should return cached query results first while rerunning the query against the backend.
        // After the backend response has returned, the query results are updated
        fetchPolicy: 'cache-and-network',
      },
    },
  });

  return client;
};
