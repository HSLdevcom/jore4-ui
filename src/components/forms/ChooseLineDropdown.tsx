import React from 'react';
import { useTranslation } from 'react-i18next';
import { RouteLine, useListAllLinesQuery } from '../../generated/graphql';
import {
  FormInputProps as ListboxInputProps,
  Listbox,
} from '../../uiComponents';

interface Props extends ListboxInputProps {
  testId: string;
}

const mapToOptionContent = (item: RouteLine) => (
  <span>{`${item.name_i18n} (${item.description_i18n})`}</span>
);

const mapToOption = (item: RouteLine) => ({
  key: item.line_id,
  value: item.line_id,
  render: () => mapToOptionContent(item),
});

export const ChooseLineDropdown = ({
  testId,
  value,
  onChange,
  onBlur,
}: Props): JSX.Element => {
  const { data: lines } = useListAllLinesQuery();
  const { t } = useTranslation();

  const options = lines?.route_line.map(mapToOption) || [];

  const selectedLine = lines?.route_line.find((item) => item.line_id === value);

  // if no line is selected, show "Choose line"
  const mapToButtonContent = (displayedLine?: RouteLine) => (
    <div className="w-full">
      {displayedLine
        ? `${displayedLine.name_i18n} (${displayedLine.description_i18n})`
        : t('routes.chooseLine')}
    </div>
  );

  return (
    <Listbox
      testId={testId}
      buttonContent={mapToButtonContent(selectedLine)}
      options={options}
      value={value}
      onChange={onChange}
      onBlur={onBlur}
    />
  );
};
