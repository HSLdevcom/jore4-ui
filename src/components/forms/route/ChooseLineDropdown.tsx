import React from 'react';
import { useTranslation } from 'react-i18next';
import { RouteLine, useListAllLinesQuery } from '../../../generated/graphql';
import {
  FormInputProps as ListboxInputProps,
  Listbox,
} from '../../../uiComponents';

interface Props extends ListboxInputProps {
  testId?: string;
}

const mapToOptionContent = (item: RouteLine) => (
  <span>{`${item.label} (${item.name_i18n})`}</span>
);

const mapToOption = (item: RouteLine) => ({
  key: item.line_id,
  value: item.line_id,
  render: () => mapToOptionContent(item),
});

export const ChooseLineDropdown = ({
  testId,
  value,
  onChange,
  onBlur,
}: Props): JSX.Element => {
  const { data: lines } = useListAllLinesQuery();
  const { t } = useTranslation();

  // @ts-expect-error typings are off?
  const options = lines?.route_line.map(mapToOption) || [];

  const selectedLine = lines?.route_line.find(
    (item) => item.line_id === value,
  ) as RouteLine;

  // if no line is selected, show "Choose line"
  const mapToButtonContent = (displayedLine?: RouteLine) => (
    <div className="w-full">
      {displayedLine
        ? `${displayedLine.label} (${displayedLine.name_i18n})`
        : t('routes.chooseLine')}
    </div>
  );

  return (
    <Listbox
      testId={testId}
      buttonContent={mapToButtonContent(selectedLine)}
      options={options}
      value={value}
      onChange={onChange}
      onBlur={onBlur}
    />
  );
};
