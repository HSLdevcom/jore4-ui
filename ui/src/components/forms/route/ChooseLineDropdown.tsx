import { debounce } from 'lodash';
import React, { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { RouteLine } from '../../../generated/graphql';
import { useChooseLineDropdown } from '../../../hooks';
import { MAX_DATE, MIN_DATE } from '../../../time';
import {
  Combobox,
  ComboboxEvent,
  FormInputProps as ListboxInputProps,
} from '../../../uiComponents';
import { DateRange } from '../common';

const DEBOUNCE_DELAY_MS = 300;

interface Props extends ListboxInputProps {
  testId?: string;
}

const mapToOptionContent = (item: RouteLine) => (
  <div>
    <span>{`${item.label} (${item.name_i18n.fi_FI})`}</span>
    <div className="text-sm">
      <DateRange
        startDate={item.validity_start || MIN_DATE}
        endDate={item.validity_end || MAX_DATE}
      />
    </div>
  </div>
);

const mapToOption = (item: RouteLine) => ({
  key: item.line_id,
  value: item.line_id,
  render: () => mapToOptionContent(item),
});

export const ChooseLineDropdown = ({
  testId,
  value,
  onChange,
  onBlur,
}: Props): JSX.Element => {
  const { t } = useTranslation();

  const [query, setQuery] = useState('');
  const [showButtonContent, setShowButtonContent] = useState(true);

  const lines = useChooseLineDropdown(query);

  const options = lines?.map(mapToOption) || [];

  const selectedLine = lines?.find((item) => item.line_id === value);

  const mapToButtonContent = (displayedLine?: RouteLine) => {
    // If no line is selected, show "Choose line"
    return (
      <div className="w-full">
        {displayedLine
          ? `${displayedLine?.label} (${displayedLine?.name_i18n.fi_FI})`
          : t('routes.chooseLine')}
      </div>
    );
  };

  const debouncedSetQuery = debounce((str) => setQuery(str), DEBOUNCE_DELAY_MS);

  const onQueryChange = (str: string) => {
    // If there is a searchword, do not show the buttonContent on top of input text
    if (str !== '') {
      setShowButtonContent(false);
    }
    debouncedSetQuery(str);
  };

  const onItemSelected = (e: ComboboxEvent) => {
    setQuery('');
    onChange(e);
    setShowButtonContent(true);
  };

  return (
    <Combobox
      id="choose-line-combobox"
      testId={testId}
      buttonContent={
        showButtonContent ? mapToButtonContent(selectedLine) : null
      }
      options={options}
      value={value}
      onChange={onItemSelected}
      onBlur={onBlur}
      onQueryChange={onQueryChange}
    />
  );
};
