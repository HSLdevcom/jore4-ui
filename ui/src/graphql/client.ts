import {
  ApolloClient,
  HttpLink,
  InMemoryCache,
  from,
  split,
} from '@apollo/client';
import { onError } from '@apollo/client/link/error';
import { removeTypenameFromVariables } from '@apollo/client/link/remove-typename';
import { WebSocketLink } from '@apollo/client/link/ws';
import { getMainDefinition } from '@apollo/client/utilities';
import { withScalars } from 'apollo-link-scalars';
import { IntrospectionQuery, buildClientSchema } from 'graphql';
import isString from 'lodash/isString';
import { DateTime, Duration } from 'luxon';
import introspectionResult from '../../graphql.schema.json';
import { isDateLike, parseDate } from '../time';
import { mapHttpToWs } from '../utils/url';
import { authRoleMiddleware, roleHeaderMap, userHasuraRole } from './auth';

const buildScalarMappingLink = () => {
  const typesMap = {
    // automatically (de)serializing between graphql date <-> luxon.DateTime types
    date: {
      serialize: (parsed: unknown) => {
        // if it's a luxon DateTime, serialize it to ISO date string
        if (DateTime.isDateTime(parsed)) {
          return parsed.toFormat('yyyy-LL-dd');
        }
        // otherwise (string, null, undefined) just pass it on as is
        return parsed;
      },
      parseValue: (raw: unknown) => {
        if (!isDateLike(raw)) {
          throw new Error(`Invalid graphql date input: '${raw}'`);
        }

        return parseDate(raw);
      },
    },
    // automatically (de)serializing between graphql interval <-> luxon.Duration types
    interval: {
      serialize: (parsed: unknown) => {
        // if it's a luxon Duration, serialize it to ISO string
        if (Duration.isDuration(parsed)) {
          return parsed.toISO();
        }
        // otherwise (string, null, undefined) just pass it on as is
        return parsed;
      },
      parseValue: (raw: unknown) => {
        if (!isString(raw)) {
          throw new Error(`Invalid graphql time input: '${raw}'`);
        }

        return Duration.fromISO(raw);
      },
    },
  };

  const schema = buildClientSchema(
    // graphql-codegen's schema typings are off...
    introspectionResult as unknown as IntrospectionQuery,
  );

  return withScalars({ schema, typesMap });
};

const getGraphqlUrl = (isTesting: boolean, isWebsocket: boolean) => {
  const path = '/api/graphql/v1/graphql';
  if (isTesting) {
    return `http://127.0.0.1:3300${path}`;
  }
  if (isWebsocket) {
    return mapHttpToWs(`${window.location.origin}${path}`);
  }
  return path;
};

const buildWebSocketLink = () => {
  return new WebSocketLink({
    // WebSocketLink doesn't work with relative url's, so we have to
    // turn relative url into absolute.
    uri: getGraphqlUrl(false, true),
    options: {
      reconnect: true,
      // TODO: deal with authentication properly. Some possibly useful info here: https://github.com/apollographql/apollo-client/issues/3967
      connectionParams: {
        headers: roleHeaderMap(userHasuraRole),
      },
    },
  });
};

// To temporarily deal with the site giving pages without any information due to stop & start
// dependencies while developing, we can catch the access-denied error and reload the page which
// triggers userinfo check, and if user is not logged -> redirect to front page and inform
// user to log in. This can and probably should be removed after we get different user accessess.
const errorLink = onError(({ graphQLErrors }) => {
  if (graphQLErrors)
    graphQLErrors.forEach(({ extensions }) => {
      if (extensions.code === 'access-denied') {
        window.location.reload();
      }
    });
});

const buildHttpLink = (isTesting: boolean) => {
  const defaultConfig = {
    uri: getGraphqlUrl(isTesting, false),
  };

  const httpLinkConfig = isTesting
    ? {
        ...defaultConfig,
        // `fetch` is not available in test context (with jest) so we have to provide it
        // eslint-disable-next-line global-require, import/no-extraneous-dependencies
        fetch: require('cross-fetch'),
        // logging in directly as admin, only works in e2e environment
        headers: { 'x-hasura-admin-secret': 'hasura' },
      }
    : defaultConfig;
  return new HttpLink(httpLinkConfig);
};

const buildConnectionLink = (isBrowser: boolean, isTesting: boolean) => {
  // because next.js might run this on server-side and websockets aren't
  // supported there, we have to check if we are on browser before
  // initializing WebSocket link
  const httpLink = buildHttpLink(isTesting);
  if (isTesting || !isBrowser) {
    return httpLink;
  }
  const wsLink = buildWebSocketLink();
  return split(
    // if running a subscription query, prefer to use wsLink
    ({ query }) => {
      const definition = getMainDefinition(query);
      const isSubscription =
        definition.kind === 'OperationDefinition' &&
        definition.operation === 'subscription';
      return isSubscription;
    },
    wsLink,
    httpLink,
  );
};

const buildCacheDefinition = () => {
  const cacheDefinition = new InMemoryCache({
    typePolicies: {
      route_line: {
        keyFields: ['line_id'],
      },
      route_route: {
        keyFields: ['route_id'],
      },
      service_pattern_scheduled_stop_point: {
        keyFields: ['scheduled_stop_point_id'],
        fields: {
          scheduled_stop_point_in_journey_patterns: {
            merge(existing, incoming) {
              // Just return incoming contents without trying to do any merging
              // to silence console warning about defining custom merge function when
              // editing route geometry.
              // The default merge function works like this as well, but for some reason
              // defining it here silences the warning.
              // "By default, the field's existing array is completely replaced by the incoming array."
              // https://www.apollographql.com/docs/react/caching/cache-field-behavior/#merging-arrays
              return incoming;
            },
          },
        },
      },
      journey_pattern_scheduled_stop_point_in_journey_pattern: {
        keyFields: ['journey_pattern_id', 'scheduled_stop_point_sequence'],
      },
      journey_pattern_journey_pattern: {
        keyFields: ['journey_pattern_id'],
        fields: {
          scheduled_stop_point_in_journey_patterns: {
            merge(existing, incoming) {
              // Just return incoming contents without trying to do any merging
              // to silence console warning about defining custom merge function when
              // editing route geometry.
              // The default merge function works like this as well, but for some reason
              // defining it here silences the warning.
              // "By default, the field's existing array is completely replaced by the incoming array."
              // https://www.apollographql.com/docs/react/caching/cache-field-behavior/#merging-arrays
              return incoming;
            },
          },
        },
      },
      route_infrastructure_link_along_route: {
        keyFields: ['route_id', 'infrastructure_link_sequence'],
      },
      infrastructure_network_infrastructure_link: {
        keyFields: ['infrastructure_link_id'],
      },
      timetables_journey_pattern_journey_pattern_ref: {
        keyFields: ['journey_pattern_ref_id'],
      },
      timetables_service_pattern_scheduled_stop_point_in_journey_pattern_ref: {
        keyFields: ['scheduled_stop_point_in_journey_pattern_ref_id'],
      },
      timetables_service_calendar_day_type: {
        keyFields: ['day_type_id'],
      },
      timetables_vehicle_schedule_vehicle_schedule_frame: {
        keyFields: ['vehicle_schedule_frame_id'],
      },
      timetables_vehicle_service_vehicle_service: {
        keyFields: ['vehicle_service_id'],
      },
      timetables_vehicle_service_block: {
        keyFields: ['block_id'],
      },
      timetables_vehicle_journey_vehicle_journey: {
        keyFields: ['vehicle_journey_id'],
      },
      timetables_timetables_query: {
        // No identifying fields here, result is cached based on query parameters.
        keyFields: [],
        fields: {
          timetables_service_calendar_substitute_operating_period: {
            merge(existing, incoming) {
              // Just return incoming contents without trying to do any merging
              // to silence console warning about defining custom merge function
              return incoming;
            },
          },
          timetables_vehicle_service_get_timetable_versions_by_journey_pattern_ids:
            {
              merge(_, incoming = []) {
                return incoming;
              },
            },
        },
      },
      timetables_passing_times_timetabled_passing_time: {
        keyFields: ['timetabled_passing_time_id'],
      },
      timing_pattern_timing_place: {
        keyFields: ['timing_place_id'],
      },
      timetables_vehicle_type_vehicle_type: {
        keyFields: ['vehicle_type_id'],
      },
      timetables_service_calendar_substitute_day_by_day_type: {
        keyFields: ['substitute_operating_day_by_line_type_id'],
      },
      timetables_service_calendar_substitute_operating_period: {
        keyFields: ['substitute_operating_period_id'],
        fields: {
          substitute_operating_day_by_line_types: {
            merge(existing, incoming) {
              // Just return incoming contents without trying to do any merging
              // to silence console warning about defining custom merge function
              return incoming;
            },
          },
        },
      },
    },
  });

  return cacheDefinition;
};

export const createGraphqlClient = () => {
  // jest and most testing frameworks set NODE_ENV to 'test' automatically
  const isTesting = process.env.NODE_ENV === 'test';

  const scalarMappingLink = buildScalarMappingLink();
  const connectionLink = buildConnectionLink(!!process.browser, isTesting);
  const removeTypenameLink = removeTypenameFromVariables();

  const link = from([
    removeTypenameLink,
    errorLink,
    scalarMappingLink,
    authRoleMiddleware,
    connectionLink,
  ]);

  const cache = buildCacheDefinition();

  const client = new ApolloClient({
    link,
    cache,
    defaultOptions: {
      watchQuery: {
        // Should return cached query results first while rerunning the query against the backend.
        // After the backend response has returned, the query results are updated
        fetchPolicy: 'cache-and-network',
      },
    },
  });

  return client;
};
