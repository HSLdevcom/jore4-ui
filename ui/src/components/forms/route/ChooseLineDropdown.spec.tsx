import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { buildLocalizedString } from '@hsl/jore4-test-db-manager';
import { act, screen } from '@testing-library/react';
import { DateTime } from 'luxon';
import {
  GetLinesForComboboxDocument,
  GetSelectedLineDetailsByIdDocument,
} from '../../../generated/graphql';
import {
  fireFullMouseClickSequence,
  render,
  sleep,
} from '../../../utils/test-utils';
import { ChooseLineDropdown } from './ChooseLineDropdown';

describe('<ChooseLineDropdown />', () => {
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const onChange = () => {};
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const onBlur = () => {};
  const value = 'line1';
  const testId = 'ChooseLineDropdown1';
  const buttonTestId = `${testId}::button`;

  const mocks: MockedResponse[] = [
    {
      request: {
        query: GetLinesForComboboxDocument,
        variables: { labelPattern: '%', date: DateTime.now().toISO() },
      },
      result: {
        data: {
          route_line: [
            {
              __typename: 'route_line',
              line_id: 'line1',
              label: '1',
              name_i18n: buildLocalizedString('Line1 name'),
              validity_start: DateTime.fromISO('2017-02-13T12:51:48.000Z'),
              validity_end: null,
            },
            {
              __typename: 'route_line',
              line_id: 'line2',
              label: '2',
              name_i18n: buildLocalizedString('Line2 name'),
              validity_start: DateTime.fromISO('2017-02-13T12:51:48.000Z'),
              validity_end: null,
            },
          ],
        },
      },
    },
    {
      request: {
        query: GetSelectedLineDetailsByIdDocument,
        variables: { line_id: 'line1' },
      },
      result: {
        data: {
          route_line_by_pk: {
            __typename: 'route_line',
            line_id: 'line1',
            label: '1',
            name_i18n: buildLocalizedString('Line1 name'),
            validity_start: DateTime.fromISO('2017-02-13T12:51:48.000Z'),
            validity_end: null,
          },
        },
      },
    },
  ];

  test('Shows correct texts when loading with preselected line', async () => {
    const { asFragment } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ChooseLineDropdown
          testId={testId}
          value={value}
          onChange={onChange}
          onBlur={onBlur}
        />
      </MockedProvider>,
    );

    // dropdown is collapsed, value is not yet bound to dropdown, 'Choose line' text shows
    expect(screen.getByTestId(testId)).toHaveTextContent('Valitse linja');
    expect(screen.queryAllByRole('option')).toHaveLength(0);
    expect(asFragment()).toMatchSnapshot();

    // wait for the graphql call to execute
    await act(() => sleep(0));

    // dropdown is collapsed, data is loaded and the pre-selected line1 is shown on the button
    expect(screen.getByTestId(testId)).toHaveTextContent('1 (Line1 name)');

    expect(asFragment()).toMatchSnapshot();
  });

  test('Shows correct texts when loading without pre-selected line', async () => {
    const { asFragment } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ChooseLineDropdown
          testId={testId}
          value={undefined}
          onChange={onChange}
          onBlur={onBlur}
        />
      </MockedProvider>,
    );

    // dropdown is collapsed, value is not yet bound to dropdown, 'Choose line' text shows
    expect(screen.getByTestId(testId)).toHaveTextContent('Valitse linja');
    expect(screen.queryAllByRole('option')).toHaveLength(0);
    expect(asFragment()).toMatchSnapshot();

    // wait for the graphql call to execute
    await act(() => sleep(0));

    // dropdown is collapsed, data is loaded and there is no pre-selected line
    expect(screen.getByTestId(testId)).toHaveTextContent('Valitse linja');

    expect(asFragment()).toMatchSnapshot();
  });

  test('Opens dropdown when clicked and shows all lines', async () => {
    const { asFragment } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ChooseLineDropdown
          testId={testId}
          value={value}
          onChange={onChange}
          onBlur={onBlur}
        />
      </MockedProvider>,
    );

    // wait for the graphql call to execute
    await act(() => sleep(0));

    // click dropdown to open it:
    const openDropdownButton = screen.getByTestId(buttonTestId);
    fireFullMouseClickSequence(openDropdownButton);

    // dropdown is open, both lines show
    expect(screen.queryByText('Valitse linja')).toBeNull();

    const items = screen.queryAllByRole('option');

    expect(items.length).toBe(2);
    expect(items[0]).toHaveTextContent('1 (Line1 name)13.2.2017 - 31.12.2050');
    expect(items[1]).toHaveTextContent('2 (Line2 name)13.2.2017 - 31.12.2050');

    expect(asFragment()).toMatchSnapshot();
  });
});
