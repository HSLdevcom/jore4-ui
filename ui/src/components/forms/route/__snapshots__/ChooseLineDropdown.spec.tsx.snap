// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChooseLineDropdown /> Opens dropdown when clicked and shows all lines 1`] = `
<DocumentFragment>
  <div
    class="relative w-full"
    data-testid="ChooseLineDropdown1"
    id="choose-line-combobox"
  >
    <input
      aria-activedescendant="headlessui-combobox-option-:r8:"
      aria-controls="headlessui-combobox-options-:r7:"
      aria-expanded="true"
      class="relative h-full w-full rounded-md border border-grey bg-white py-3 px-2"
      data-testid="ChooseLineDropdown1::input"
      id="headlessui-combobox-input-:r4:"
      role="combobox"
      type="text"
    />
    <button
      aria-controls="headlessui-combobox-options-:r7:"
      aria-expanded="true"
      aria-haspopup="true"
      class="absolute inset-y-0 right-0 flex h-full w-full items-center justify-end px-3 text-left focus:outline-none"
      data-testid="ChooseLineDropdown1::button"
      id="headlessui-combobox-button-:r5:"
      tabindex="-1"
      type="button"
    >
      <div
        class="w-full"
      >
        1 (Line1 name)
      </div>
      <svg
        class="mr-1 ml-auto text-2xl text-tweaked-brand"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <i
        class="icon-arrow text-tweaked-brand transition duration-150 ease-in-out -rotate-180"
        style="font-size: 10px;"
      />
    </button>
    <ul
      aria-activedescendant="headlessui-combobox-option-:r8:"
      aria-labelledby="headlessui-combobox-button-:r5:"
      class="absolute left-0 z-10 w-full rounded-b-md border border-black border-opacity-20 bg-white shadow-md focus:outline-none transition ease-out duration-100 opacity-0 scale-95"
      id="headlessui-combobox-options-:r7:"
      role="listbox"
    >
      <li
        aria-selected="true"
        id="headlessui-combobox-option-:r8:"
        role="option"
        tabindex="-1"
      >
        <span
          class="bg-background flex group text-left pr-2 py-2 focus:outline-none items-center"
        >
          <svg
            class="mx-1 text-2xl text-black"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
            />
          </svg>
          <div>
            <span>
              1 (Line1 name)
            </span>
            <div
              class="text-sm"
            >
              13.2.2017 - 31.12.2050
            </div>
          </div>
        </span>
      </li>
      <li
        id="headlessui-combobox-option-:r9:"
        role="option"
        tabindex="-1"
      >
        <span
          class=" flex group text-left pr-2 py-2 focus:outline-none items-center"
        >
          <svg
            class="mx-1 text-2xl invisible"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
            />
          </svg>
          <div>
            <span>
              2 (Line2 name)
            </span>
            <div
              class="text-sm"
            >
              13.2.2017 - 31.12.2050
            </div>
          </div>
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`<ChooseLineDropdown /> Shows correct texts when loading with preselected line 1`] = `
<DocumentFragment>
  <div
    class="relative w-full"
    data-testid="ChooseLineDropdown1"
    id="choose-line-combobox"
  >
    <input
      aria-expanded="false"
      class="relative h-full w-full rounded-md border border-grey bg-white py-3 px-2"
      data-testid="ChooseLineDropdown1::input"
      id="headlessui-combobox-input-:r0:"
      role="combobox"
      type="text"
    />
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="absolute inset-y-0 right-0 flex h-full w-full items-center justify-end px-3 text-left focus:outline-none"
      data-testid="ChooseLineDropdown1::button"
      id="headlessui-combobox-button-:r1:"
      tabindex="-1"
      type="button"
    >
      <div
        class="w-full"
      >
        Valitse linja
      </div>
      <svg
        class="mr-1 ml-auto text-2xl text-tweaked-brand"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <i
        class="icon-arrow text-tweaked-brand transition duration-150 ease-in-out rotate-0"
        style="font-size: 10px;"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`<ChooseLineDropdown /> Shows correct texts when loading with preselected line 2`] = `
<DocumentFragment>
  <div
    class="relative w-full"
    data-testid="ChooseLineDropdown1"
    id="choose-line-combobox"
  >
    <input
      aria-expanded="false"
      class="relative h-full w-full rounded-md border border-grey bg-white py-3 px-2"
      data-testid="ChooseLineDropdown1::input"
      id="headlessui-combobox-input-:r0:"
      role="combobox"
      type="text"
    />
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="absolute inset-y-0 right-0 flex h-full w-full items-center justify-end px-3 text-left focus:outline-none"
      data-testid="ChooseLineDropdown1::button"
      id="headlessui-combobox-button-:r1:"
      tabindex="-1"
      type="button"
    >
      <div
        class="w-full"
      >
        1 (Line1 name)
      </div>
      <svg
        class="mr-1 ml-auto text-2xl text-tweaked-brand"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <i
        class="icon-arrow text-tweaked-brand transition duration-150 ease-in-out rotate-0"
        style="font-size: 10px;"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`<ChooseLineDropdown /> Shows correct texts when loading without pre-selected line 1`] = `
<DocumentFragment>
  <div
    class="relative w-full"
    data-testid="ChooseLineDropdown1"
    id="choose-line-combobox"
  >
    <input
      aria-expanded="false"
      class="relative h-full w-full rounded-md border border-grey bg-white py-3 px-2"
      data-testid="ChooseLineDropdown1::input"
      id="headlessui-combobox-input-:r2:"
      role="combobox"
      type="text"
    />
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="absolute inset-y-0 right-0 flex h-full w-full items-center justify-end px-3 text-left focus:outline-none"
      data-testid="ChooseLineDropdown1::button"
      id="headlessui-combobox-button-:r3:"
      tabindex="-1"
      type="button"
    >
      <div
        class="w-full"
      >
        Valitse linja
      </div>
      <svg
        class="mr-1 ml-auto text-2xl text-tweaked-brand"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <i
        class="icon-arrow text-tweaked-brand transition duration-150 ease-in-out rotate-0"
        style="font-size: 10px;"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`<ChooseLineDropdown /> Shows correct texts when loading without pre-selected line 2`] = `
<DocumentFragment>
  <div
    class="relative w-full"
    data-testid="ChooseLineDropdown1"
    id="choose-line-combobox"
  >
    <input
      aria-expanded="false"
      class="relative h-full w-full rounded-md border border-grey bg-white py-3 px-2"
      data-testid="ChooseLineDropdown1::input"
      id="headlessui-combobox-input-:r2:"
      role="combobox"
      type="text"
    />
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="absolute inset-y-0 right-0 flex h-full w-full items-center justify-end px-3 text-left focus:outline-none"
      data-testid="ChooseLineDropdown1::button"
      id="headlessui-combobox-button-:r3:"
      tabindex="-1"
      type="button"
    >
      <div
        class="w-full"
      >
        Valitse linja
      </div>
      <svg
        class="mr-1 ml-auto text-2xl text-tweaked-brand"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <i
        class="icon-arrow text-tweaked-brand transition duration-150 ease-in-out rotate-0"
        style="font-size: 10px;"
      />
    </button>
  </div>
</DocumentFragment>
`;
